### PR TITLE
Fix cdfast convergence

### LINKED ISSUE #275: failing comparison of elastic net with scikit-learn
I tried to compare pyglmnet.GLM with ElasticNet from scikit-learn and could not get it work. Code to reproduce:

```python
import numpy as np
from sklearn.datasets.samples_generator import make_regression
from sklearn.linear_model import ElasticNet
from pyglmnet import GLM

def rmse(a, b):
    return np.sqrt(np.mean((a - b) ** 2))

X, Y, coef_ = make_regression(
    n_samples=1000, n_features=1000,
    noise=0.1, n_informative=10, coef=True,
    random_state=42)

alpha = 0.1
l1_ratio=0.5

sk = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, tol = 1e-5).fit(X, Y)
pg = GLM(distr='gaussian', alpha=l1_ratio, reg_lambda=alpha, solver='cdfast', tol = 1e-5).fit(X, Y)
print('in-sample rmse sklearn = {}, pyglmnet = {}'.format(rmse(Y, sk.predict(X)), rmse(Y, pg.predict(X))))
```

Result:
in-sample rmse sklearn = 12.756054997216442, pyglmnet = 161.03496460055504

### PR DESCRIPTION
**Pull request by peterfoley605**
fixes #275 

changes test_glmnet to include tests with with nonzero lambda. Changes `GLM(solver='cdfast').fit()` execution path to recalculate `z` each time since it is actually invalidated on each iteration.

To speed up my testing I also added a manifest to make the package directly pip-installable, which is likely a useful addition even though it's not strictly related to the fix.

### CODE REVIEW

File: tests/test_pyglmnet.py [hunk 1]
```diff
@@ -164,61 +164,91 @@ def test_glmnet():
...
+        for reg_lambda in reg_lambdas:
+            betas_ = list()
+            for solver in solvers:
+
+                np.random.seed(random_state)
+
+                X_train = np.random.normal(0.0, 1.0, [n_samples, n_features])
+                y_train = simulate_glm(distr, beta0, beta, X_train,
+                                       sample=False)
+                alpha = 0.5
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: was this change intentional?

File: tests/test_pyglmnet.py [hunk 2]
```diff
@@ -164,61 +164,91 @@ def test_glmnet():
...
+                          reg_lambda=reg_lambda, tol=1e-3, max_iter=5000,
+                          alpha=alpha, solver=solver,
+                          score_metric=score_metric,
+                          random_state=random_state, callback=callback)
+                assert(repr(glm))
+
+                glm.fit(X_train, y_train)
+
+                # verify loss decreases or increases only slightly
+                loss_trace_decreased = np.diff(loss_trace) <= 1e-3
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: I suppose an easy check (that would hold for any reg_lambda) would be to count the number of iterations it's non-zero and to assert that it's not more than 2 or so.

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -731,6 +731,7 @@ def fit(self, X, y):
 
             elif self.solver == 'cdfast':
                 beta_old = deepcopy(beta)
+                z = beta[0] + np.dot(X, beta[1:])
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: this seems more like a hack, no? The cache that is computed before is useless now. Could it be fixed instead of recomputing this here?

Comment (peterfoley605) [reply pyglmnet/pyglmnet.py]: it's recalculated in this block of code since it's invalidated here and it's needed by the call to _cdfast. Since it's the only place that _cdfast is called, though, I'll go ahead and remove `z` from the _cdfast interface entirely.

File: tests/test_pyglmnet.py (hunk 1 above)
Comment (peterfoley605) [reply tests/test_pyglmnet.py]: yes, so that issues related to either L1 or L2 norms could break the test. Ex. prox_ uses threshold `reg_lambda * alpha`, so if alpha is zero, problems that arise with nonzero alpha will be hidden similarly to what this PR is fixing for `reg_lambda=0.0`

File: tests/test_pyglmnet.py [hunk 4]
```diff
@@ -164,61 +164,91 @@ def test_glmnet():
...
+                                   reg_lambda, X_train, y_train, 2.0, None,
+                                   np.concatenate(([beta0], beta)))
+                    assert_allclose(loss_trace[-1], l_true,
+                                    rtol=1e-4, atol=1e-5,
+                                    err_msg=('Final loss trace value different'
+                                             ' from true loss '
+                                             ' on distr={d} solver={s}'
+                                             ' with reg_lambda={rl}'
+                                             ).format(d=distr, s=solver,
+                                                      rl=reg_lambda))
```
[the remark this answers is not in the document]
Comment (peterfoley605) [reply tests/test_pyglmnet.py]: Those are there because it made it a ton easier for me to solve this problem. I'd recommend keeping them so that the problematic distr/solver is immediately clear without manual pdb work.

File: MANIFEST.in [hunk 5]
```diff
@@ -0,0 +1,5 @@
+include *.rst *.md
+include LICENSE
+recursive-include doc *
```
[the remark this answers is not in the document]
Comment (peterfoley605) [reply MANIFEST.in]: I'll go ahead and just drop MANIFEST.in from this PR since it's tangential and I see there's an existing issue to solve it properly.

**Comment by peterfoley605** [comment]
re. flake8 errors, I addressed everything that's touched by this PR (using `git diff main/master -U0  | flake8 --diff -`). The other flake8 issues are leftovers from `master`.

File: tests/test_pyglmnet.py (hunk 2 above)
Comment (peterfoley605) [reply tests/test_pyglmnet.py]: I'm not sure what makes the most sense, since I'm not extremely familiar with how cdfast should behave. Glancing at loss traces, it seems like increases in loss are not uncommon within cdfast (though when they appear, they're small and usually resolved by the next iteration), in which case a total count would throw false positives when max_iter is high.

Do you think tracking a count of sequential steps where loss increases would make sense? Ex. xgboost's early_stopping_rounds argument where it stops if the loss (on validation) doesn't decrease for that many iterations in a row.

Here since repeated in-bag loss increases signal a bug, I'd think we'd want to monitor those increases and throw an exception within `fit()` itself, which would be natural to integrate into #277 .

File: tests/test_pyglmnet.py [hunk 6]
```diff
@@ -164,61 +165,93 @@ def test_glmnet():
...
+                loss_increase_runlengths = np.array(loss_increase_runlengths)
+                assert np.all(loss_increase_runlengths <= 5), \
+                    ('Loss increased for {maxlen} consecutive iterations'
+                     ' on distr={d} solver={s} with'
+                     ' reg_lambda={rl}'
+                     ).format(d=distr, s=solver,
+                              rl=reg_lambda,
+                              maxlen=np.max(loss_increase_runlengths))
+
+                if reg_lambda == 0.0:
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: don't like the 'if' condition after the loop over reg_lambda. We are at 4 levels of nesting now. Could this be avoided? Maybe just test for reg_lambda = 0.1 would simplify things?

File: tests/test_pyglmnet.py (hunk 2 above)
Comment (jasmainak) [reply tests/test_pyglmnet.py]: I think the loss can go up in cdfast but it should not go up in batch (at least for small step sizes). 

early_stopping_rounds sounds like a good idea. However, it might be good to check what's done in the R glmnet package or statsmodel. We may not want to commit too much of the API for one particular solver (cdfast in this case). I know that sklearn uses coordinate descent with cython and in that case you don't have to worry about losses going up.

File: setup.cfg [hunk 7]
```diff
@@ -7,7 +7,11 @@ ignore-files = (benchmarks.py|datasets.py)
 
 [flake8]
 exclude = __init__.py,six.py,funcsigs.py
-ignore = E241
+ignore =
+  # W504: Line break occurred after a binary operator
+  W504,
```
[the remark this answers is not in the document]
Comment (jasmainak) [reply setup.cfg]: ok I [saw this](https://github.com/glm-tools/pyglmnet/pull/278#issuecomment-470179287) late. Maybe let's leave it as it is. I'm resolving the comment.

File: tests/test_pyglmnet.py [hunk 8]
```diff
@@ -164,61 +165,93 @@ def test_glmnet():
...
+
+                glm.fit(X_train, y_train)
+
+                # verify that loss doesn't increase more than tolerance
+                # for too many consecutive iterations
+                loss_increased = (np.diff(loss_trace) > glm.tol)
+                loss_increase_runlengths = [sum(grp) for
+                                            val, grp in
+                                            itertools.groupby(loss_increased)
+                                            if val]
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: if val is True? It's better to be explicit

**Comment by jasmainak** [comment]
@peterfoley do you mind if I push a commit to conform to the coding style so we can merge this?

**Comment by peterfoley605** [comment]
@jasmainak I just enabled "allow edits from maintainers" so you can push commits.

**Comment by peterfoley605** [comment]
@jasmainak I massively pared down this PR since the changes other than cdfast caching were resolved by other PRs over the past few months. Could you re-review?

As a reminder for why `z` from `_cdfast` shouldn't be cached outside `_cdfast` itself -- immediately after `_cdfast` is called by `fit()`, `beta` is updated by `self._prox`, which makes the cached `z` inconsistent with `beta`.

**Comment by jasmainak** [comment]
@peterfoley thanks for getting back to this. I pushed a test. There is one problem though. On master, if I do:

```sh
$ py.test --cov=pyglmnet tests/ --verbose --durations=10
```

I get:

```sh
========================== slowest 10 test durations ===========================
9.38s call     tests/test_pyglmnet.py::test_random_state_consistency
9.19s call     tests/test_pyglmnet.py::test_cv
8.62s call     tests/test_pyglmnet.py::test_glmcv[probit]
2.77s call     tests/test_pyglmnet.py::test_glmcv[poisson]
2.12s call     tests/test_pyglmnet.py::test_glmcv[softplus]
1.85s call     tests/test_pyglmnet.py::test_glmcv[binomial]
1.78s call     tests/test_pyglmnet.py::test_glmcv[gamma]
1.66s call     tests/test_pyglmnet.py::test_glmcv[gaussian]
0.37s call     tests/test_pyglmnet.py::test_fetch_datasets
0.31s call     tests/test_metrics.py::test_deviance
```

On this branch, I get:

```sh
=========================== slowest 10 test durations ============================
65.17s call     tests/test_pyglmnet.py::test_glmcv[probit]
18.17s call     tests/test_pyglmnet.py::test_glmcv[poisson]
9.96s call     tests/test_pyglmnet.py::test_glmcv[softplus]
9.26s call     tests/test_pyglmnet.py::test_random_state_consistency
9.00s call     tests/test_pyglmnet.py::test_glmcv[binomial]
8.48s call     tests/test_pyglmnet.py::test_cv
7.54s call     tests/test_pyglmnet.py::test_glmcv[gaussian]
3.67s call     tests/test_pyglmnet.py::test_fetch_datasets
1.91s call     tests/test_pyglmnet.py::test_glmcv[gamma]
1.78s call     tests/test_pyglmnet.py::test_compare_sklearn
```

It's not a proper benchmark but it seems a big slowdown, no?

**Comment by jasmainak** [comment]
By the way, have you looked at this: https://github.com/glm-tools/pyglmnet/issues/289 ?

**Comment by jasmainak** [comment]
Regardless of my comment above about the slowdown, I think your changes are technically correct @peterfoley605 . I think the slowdown might be related to the stopping criteria? 

If you can update the `whats_new.rst` with a clear explanation of what you did, I would be happy to merge this PR now.

**Comment by jasmainak** [comment]
Also, just to be sure I understand correctly, there is still caching within one cycle ... just that there is no caching from one cycle of Newton updates to the next. That seems to me to be the correct thing to do.

**Comment by peterfoley605** [comment]
The large pytest slowdown is surprising given that it adds only one relatively simple linear algebra operation per iteration (you're correct that it still caches within _cdfast).

I replicated the pytest slowdown on my machine, but I haven't been able to replicate it outside pytest by running the contents of `test_glmnet` in a standalone script. It seems like a pytest-specific thing, perhaps related to GC or memory tracing, since `z`'s memory is now allocated with each iteration?

I'll go ahead with the `whats_new.rst` update and we can get this merged.

**Comment by jasmainak** [comment]
I think it's a question of how fast the algorithm converges rather than the number of flops. We'd need to plot convergence curves to nail it down.

But anyway, I merged it because it fixes an important problem. Thanks for the contribution!

**Comment by jasmainak** [comment]
> running the contents of test_glmnet in a standalone script

could you share this script on a gist or something?

**Comment by peterfoley605** [comment]
@jasmainak I moved the test into a single ipython notebook that creates fresh conda environments and tests two commits: https://gist.github.com/peterfoley605/53d6faa0c491ffbc8a682ff147d1d6e0
or if that's not rendering:
https://nbviewer.jupyter.org/gist/peterfoley605/53d6faa0c491ffbc8a682ff147d1d6e0

**Comment by jasmainak** [comment]
@peterfoley sorry in the delay getting back here. I could not open your notebook locally on my computer. And the nbviewer link doesn't seem to work either ...
